### ISSUE TITLE
[Bug]: Passing http://localhost:4201 as the URI in SIWE message on a localhost:4201 site seems to close MM popup without any error thrown

### ISSUE THREAD

**Issue by bschorchit** [MetaMask/metamask-extension#18188]
### Describe the bug

This was originally reported to us [here on Slack](https://consensys.slack.com/archives/G8RSKCNCD/p1678906632979459) and here https://github.com/MetaMask/metamask-extension/issues/17707#issuecomment-1470728017

### Steps to reproduce

Here is a example of message to sign that was reported to cause this while on localhost : https://github.com/MetaMask/metamask-extension/issues/17707#issuecomment-1470786260

### Version

10.26.1

### Build type

None

### Browser

Other (please elaborate in the "Additional Context" section)

### Operating system

Other (please elaborate in the "Additional Context" section)

### Additional context

Not sure on browser and OS info.

**Comment by bschorchit** [comment]
Update from @seaona :
after trying several times, one time the popup didn't open - no error neither. I could also see SIWE screen too in other trials and another thing I see is that on the URI it's set https://localhost:9011/ but I am using http.

![image (8)](https://user-images.githubusercontent.com/34968356/225614748-df438a5d-c93c-4164-ae6d-22e4bc188e28.png)

**Comment by yonigoldberg** [comment]
@bschorchit is there an ETA on this issue?

**Comment by yonigoldberg** [comment]
@bschorchit - I opened a pr for a fix

**Comment by bschorchit** [comment]
Thanks for that, @yonigoldberg ! We rolled out a hotfix (v10.26.2) yesterday. Is the current behavior on 10.26.2 on localhost still an issue for you?

**Comment by yonigoldberg** [comment]
@digiwand @jpuri @danjm - 
This is related to your fix in: https://github.com/MetaMask/metamask-extension/pull/18200
Do you mind reviewing https://github.com/MetaMask/metamask-extension/pull/18296?

**Comment by yonigoldberg** [comment]
Thanks @digiwand, I added the test that you mentioned [here](https://github.com/MetaMask/metamask-extension/pull/18296#pullrequestreview-1357929773)  (I can't comment on the PR due to lack of permissions).

**Comment by naugtur** [comment]
@yonigoldberg I don't have full context so I'm making assumptions from my webapp security background. This might be wrong. That said, I think port being optional might mean it defaults to 80?

I'll take a closer look. Or maybe someone more familiar with the Origin guarantees in SIWE can chime in.   
My participation is just the kneejerk reflex when I see protocol/port being omited on something called origin.

**Comment by digiwand** [comment]
we can loosen restrictions later if we see a use-case for ports on remote origins. For now, unblocking ports on localhost seems useful and matches the issue case. Should remote origins with mismatched ports present themselves, they will not be blocked and only be presented with a warning.

curious to learn more about the port use cases here and why this may or may not be a concern outside of localhost.

**Comment by naugtur** [comment]
ERC-4361, by referencing `authority` from RFC-3986 is implicitly dragging the concept of `URI scheme` along with it as `authority` is defined within the scope of a `URI scheme`.

For the purpose of RFC-3986 `authority` used in `domain` of ERC-4361, how is the `URI scheme` defined?   
It is the `URI scheme` that defines the default port value, so it's necessary for the interpretation of the domain that doesn't contain the `:port`.  
eg. If we assume `http` as the URI scheme, lack of `:port` can be interpreted to mean `:80` or empty. the RFC-3986 itself is not clear about it and we'd have to reach for the http spec, where (I assume from prior experience) the interpretation ends up being `:80` not empty.

ERC-4361 would have to define `domain` as RFC-3986 `authority` with a `URI scheme` that doesn't specify a default port to allow ignoring the port under all circumstances.

I think the lack of a field defining the `URI scheme` in ERC-4361 spec may be considered as it being implicitly declared void. The domain is coming from a http or https scheme though. 

also, ERC-4361 "Out of scope" section says:  
>  The specific mechanisms to ensure domain-binding.

In the current shape of ERC-4361 the burden of defining what `URI scheme` to consider for checking that the `domain` matches the actual signing request source falls on said `specific mechanisms`.

Unless ERC-4361 is updated to explicitly define `URI scheme` for `authority` used as `domain`, it's on us to decide how to interpret the missing port.

Specifically, regarding my earlier comments:
- sameorigin-policy is not intended to be enforced here, but can serve as a guideline in determining what to consider different
- It's on us, not the ERC, to decide what rules we want to use for domain matching and making a carve-out for localhost is valid too.
- If we have access to protocol (http vs https, maybe file too?) the `URI scheme` is known to us. 
- For http/https addresses I believe lack of port specified should be interpreted as default port not as a wildcard for the purpose of making the comparison, so example.com is a match for request origin example.com:443 under https but not for example.com:1337

**Comment by holantonela** [comment]
Thanks for this analysis @naugtur! I think it makes sense to use the patters that we already have in place but reflecting the heuristics presented in your table:

    ❌ - mismatch error, prevent signing ---> API returns an error, the user never get exposed to that UI
    ⚠️ - show a prominent scary warning before proceeding to signature ---> We use the warning message + the second step to confirm. 
    👍 - match, ask user for signature --> regular SIWE screen.

**Comment by naugtur** [comment]
After sleeping on it I now think the most secure option would be to:

- Enforce strict port matching by default
- Require the request origin to be https, therefore the only option for no port is default to 443
- Introduce a setting in the wallet to enable developer mode of sorts that disables those strict protections and replaces them with a warning (behavior 4)
- Why have a warning if it's a dev mode? Because it's not that hard to convince people to flip a switch in settings before proceeding with the phish.

**Comment by yonigoldberg** [comment]
@legobeat and @naugtur - great analysis of the specs. Thank you 🙏 

I am also convinced that the optional port should only be allowed when it is not provided. Otherwise, it should be enforced.   With that, I am going to close the PR.

@naugtur - one exception about enforcing https is `localhost`.  From a developer standpoint, it should be allowed to use `http` on localhost.

**Comment by digiwand** [comment]
many thanks @legobeat and @naugtur!

These examples are useful and I'd like to propose a couple more examples. I agree with most of them except I have a concern with 1A. 
(sorry, I know you said "should not need further discussion." but I'm going for it :D) @naugtur 

--- 

### 1A 

I see ❌  now under the "warn only" session. Was this meant to be ⚠️😱? 
_While we do not support a developer-mode setting_, I don't think we should block this case entirely. 

One scenario for the mismatched domain was explained in the related ticket #17707. 

from  @HeyFloM 's usecase https://github.com/MetaMask/metamask-extension/issues/17707#issuecomment-1470877080:
> On our side, this is problematic for all integrations of Metamask with Unity WebGL games/metaverse.
Why? Because all WebGL builds needs to be hosted on specific hosting services like itch.io so we can easily test them without having to upload each build on our Amazon S3 Buckets. So each development build have this error right now
image

they did figure out a solution for production, but it continues to cause friction for development: 

> After identifying the issue, we aligned the URL of the domain hosting the production Unity build with the server URL that manages the Unity SDK and Web3 authentication (Moralis Unity SDK self-hosted server). As a result, there are no longer any problems with the production part.

> However, the impact is still present on the development side because we use different hosting services for speed and agility to export and test quickly. These services make it very easy to see what are the results of your latest dev changes in the build, but they are on another domain.
We could create a test environment on our servers, but this would require our developers to:
 > - export the WebGl build from Unity
 > - Manually import it on our Amazon S3 buckets
 > - Convert it to react
 > - Align the different domain URLs manually
 > - And this at least 4 times per day, which would be a significant timeloss

----

## Additional Examples

**Situation 1D**: A page hosted at `http://evil.com` requests SIWE signature with domain set to `evil.com`.
Proposal: 1D | ❌ | ❌ | ❌ | ⚠️😱 
Purpose: Evaluate non-TLS-protected website situation 


**[maybe] Situation 1E:** A page hosted at `https://www.example.com` requests SIWE signature with domain set to `https://example.com`.
Proposal: 1E | ❌ | ❌ | ❌ | ⚠️😱 
Purpose:  This would be to track that we have considered this case of the `www` subdomain though I believe we should apply the same behavior as we do with other subdomains (Situation 1B); as `www` is another subdomain itself. Like other subdomains, it's possible for the "www" subdomain to host different resources than the root domain. (Relates to https://github.com/MetaMask/metamask-extension/issues/18332) 

----

Implementing strict mode alongside a developer mode setting sounds good to me. 

If we enforce strict mode for any of these situations, we'll need to inform developers of this change and give them time to make updates if needed.

**Comment by digiwand** [comment]
I think that makes sense. Distributing TLS certificates to the CDN will allow the appropriate domains to host those servers. Thanks for the advice @legobeat!

okay, I agree 1A no longer needs a change. I do still believe blocking the request entirely should be communicated before implementing this situation.

**Comment by naugtur** [comment]
For 1A anything other than rejection would be against the spec. It's the only thing ERC is clearly requiting us to do. Note that it's not a subdomain issue but a totally different domain.

I'm ok with allowing http for localhost regardless of how strict we decide to be.

**Comment by digiwand** [comment]
okay, that makes sense. I think my communication was confusing. I think it's a good idea to block 1A once we have a developer setting. Even more so with the setup @legobeat mentioned above.

> I'm ok with allowing http for localhost regardless of how strict we decide to be.

@naugtur when you mentioned allowing these cases, were you thinking with or without the warning modal? 

----

_Just read the [updated comment](https://github.com/MetaMask/metamask-extension/issues/18188#issuecomment-1499919118) now @legobeat and responding to 1D comment_ 

re: 1D. okay, I see I missed this earlier. I agree and updated the comment with 👍. [Snippet from EIP-4361 spec](https://eips.ethereum.org/EIPS/eip-4361): 
> The following concerns are out of scope for this version of the specification to define:
> ...
> Protocols for use without TLS connections.

Good idea to include a specific setting to disable domain-binding in Flash/beta as a plan B. I've created a new ticket here
https://github.com/MetaMask/metamask-extension/issues/18531 to implement a developer mode setting, and noted this consideration.

----

I've created 2 new tickets to support our implementation:
1. https://github.com/MetaMask/metamask-extension/issues/18529
2. https://github.com/MetaMask/metamask-extension/issues/18531



@legobeat - reviewing the related PRs today

**Comment by bschorchit** [comment]
I've just checked and https://github.com/MetaMask/metamask-extension/pull/18518 is part of 10.31 release - which should be rolling out soon (next week I think) to users.

I don't think we need to wait to ship https://github.com/MetaMask/metamask-extension/issues/18529 before closing this off. We can track and discuss anything related to planning/communicating and blocking mismatched domains directly on that issue (and other related issues) and come back to this one for reference even if closed.
